src/goto-instrument/contracts/contracts.cpp

```diff
@@ -209,7 +209,13 @@ void code_contractst::check_apply_loop_contracts(
       converter.goto_convert(old_decreases_assignment, havoc_code, mode);
     }
 
-    goto_function.body.destructive_insert(std::next(loop_head), havoc_code);
+    // Forward the loop_head iterator until the start of the body.
+    // This is necessary because complex C loop_head conditions could be
+    // converted to multiple GOTO instructions (e.g. temporaries are introduced).
+    const auto head_loc = loop_head->source_location();
+    while(loop_head->source_location() == head_loc)
```

Comment:
A. Frankly, this makes me nervous.  I don't think we have any (documented, checked, folklore, etc.) invariants about `source_location` and making it semantically meaningful.  There are a few places that do, things like the coverage code (I think @peterschrammel is the expert here) but ... I am concerned about this.

B. I can't think of a better way of distinguishing:

```
while (a && b) {
...
}
```

from

```
while (a) {
 if (! b) break;
}
```

I guess they produce the same `goto_programt`?

C. Do we need some common way of marking where the loop head / body is?  Should this code go into the `natural_loops` detection code?  Do we need a marker instruction?  When working on other analysis tools the question of "what is 'in a loop'" has come up and it is a lot more complex than it first appears.

Reply:
Sharing some background discussion in here: I have the very same concern, and hence asked for some test to be added (the one you were wondering about) just so that at least we can investigate afterwards. We will likely have to rewrite some of this.

Reply:
Having one consistently used interface for identifying loops, the exit conditions, the exit locations, the "body", etc. would be great.  Note that the syntactic and semantic definitions of loops are different.

I am happy for this to be flagged as a future issue as I think the rest of the code base is not exactly consistent on this.